### PR TITLE
Improve performance of `get_missing_sessions`

### LINKED ISSUE #2793: Unreachable servers cause poor performance in `OlmMachine::get_missing_sessions`
Suppose you have a room which includes Bob, whose server times out any federation requests.

When sending a message in the room, the application must call `OlmMachine::get_missing_sessions` (which delegates to [`SessionManager::get_missing_sessions`](https://github.com/matrix-org/matrix-rust-sdk/blob/f945a50d22aabb14e018d74303d6d6d26779d423/crates/matrix-sdk-crypto/src/session_manager/sessions.rs#L227) with all the users in the room, including Bob).

`get_missing_sessions` calls `get_user_devices` which waits up to 5 seconds for a device list to be received (it is assumed that there are attempts being made to populate the device list elsewhere). Since Bob's server is unreachable, this will never happen.

Worse, this happens **in series** for each user in the room on an unreachable server.

There is an attempt to filter out "failed" servers in `get_missing_sessions`; however it doesn't help because the only "failures" recorded here are those where a `/keys/claim` request has previously returned a "failures" entry for the given server. Since we don't know about any devices for Bob, we will never make a `/keys/claim` request for any of his devices.

### PR DESCRIPTION
**Pull request by richvdh**
Two sets of improvements to `get_missing_sessions` here:

 * Currently, for any users who have an empty device list, we call `KeyQueryManager::wait_if_user_key_query_pending`, which waits up to 5 seconds for a `/keys/query` response. (Arguably, we should be waiting longer: it is not unusual for such a request to take a while.)

   The problem is that that user's server may have been blacklisted, in which case we won't even be trying to do `/keys/query` resquests for that user, so we'll be waiting for something that never happens.

   To fix this, let's check the failures list when deciding if we should wait for a user's devices.

 * Separately, but closely related: for each user thus affected, we do the wait *in series*. This is a bit silly: there is no point waiting 50 times. We can parallelise the work.

Fixes #2793.

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/identities/manager.rs [hunk 1]
```diff
@@ -878,6 +883,125 @@ impl IdentityManager {
...
+                devices_by_user.insert(user_id, updated_devices);
+            }
+        }
+
+        Ok(devices_by_user)
+    }
+
+    /// Helper for get_user_devices_for_encryption.
+    ///
+    /// Waits for any pending `/keys/query` for the given device. If one was
```
Comment (richvdh) [inline crates/matrix-sdk-crypto/src/identities/manager.rs:983]:
```suggestion
    /// Waits for any pending `/keys/query` for the given user. If one was
```

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs [hunk 2]
```diff
@@ -247,11 +225,21 @@ impl SessionManager {
         let mut missing: BTreeMap<_, BTreeMap<_, _>> = BTreeMap::new();
         let mut timed_out: BTreeMap<_, BTreeSet<_>> = BTreeMap::new();
 
-        // Add the list of devices that the user wishes to establish sessions
-        // right now.
-        for user_id in users.filter(|u| !self.failures.contains(u.server_name())) {
-            let user_devices = Box::pin(self.get_user_devices(user_id)).await?;
+        let unfailed_users = users.filter(|u| !self.failures.contains(u.server_name()));
 
+        // XXX: You'd think this would be unnecessary, but if you omit it, Rust starts
```
Comment (jplatte) [inline crates/matrix-sdk-crypto/src/session_manager/sessions.rs:230]: What might be related is the generic `users` possibly not implementing `Send`. Maybe adding `+ Send` to the `impl Iterator` would fix the issue.

Comment (jplatte) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:230]: (as per discussion on Matrix, this does not help)

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs [hunk 3]
```diff
@@ -225,16 +225,21 @@ impl SessionManager {
...
-                self.key_request_machine
-                    .identity_manager()
-                    .get_user_devices_for_encryption(user_id),
-            )
-            .await?;
+        let unfailed_users = users.filter(|u| !self.failures.contains(u.server_name()));
 
+        // XXX: You'd think this would be unnecessary, but if you omit it, Rust starts
+        // yelling about lifetimes and Send for reasons that I don't really understand.
+        let unfailed_users: Vec<_> = unfailed_users.collect();
```
Comment (bnjbvr) [inline crates/matrix-sdk-crypto/src/session_manager/sessions.rs:232]: You can remove this if you change `impl IntoIterator` in `impl Iterator` in the signature of `get_user_devices_for_encryption`.

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Nice performance improvements. Do you have some measures on real-world use cases? (absolutely not mandatory, but I'm just curious :))

File: crates/matrix-sdk-crypto/src/session_manager/sessions.rs (hunk 3 above)
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/session_manager/sessions.rs:232]: Of course I can. Thanks for that, I was totally blanking on it.

File: crates/matrix-sdk-crypto/src/identities/manager.rs [hunk 4]
```diff
@@ -893,14 +893,23 @@ impl IdentityManager {
...
         &self,
-        user_id: &UserId,
-    ) -> StoreResult<HashMap<OwnedDeviceId, ReadOnlyDevice>> {
+        users: impl IntoIterator<Item = &UserId>,
+    ) -> StoreResult<HashMap<OwnedUserId, HashMap<OwnedDeviceId, ReadOnlyDevice>>> {
         // how long we wait for /keys/query to complete
         const KEYS_QUERY_WAIT_TIME: Duration = Duration::from_secs(5);
 
-        let user_devices = self.store.get_readonly_devices_filtered(user_id).await?;
+        // first of all, check the store for each user.
```
[the remark this answers is not in the document]
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/identities/manager.rs]: Yes, sorry. I think I got them all, but this is a bad habit of mine.

File: crates/matrix-sdk-crypto/src/identities/manager.rs [hunk 5]
```diff
@@ -912,9 +921,11 @@ impl IdentityManager {
...
         // bother.
 
-        if !user_devices.is_empty() {
-            return Ok(user_devices);
-        }
+        let users_with_no_devices: BTreeSet<OwnedUserId> = devices_by_user
+            .iter()
+            .filter(|(_user_id, devices)| devices.is_empty())
+            .map(|(user_id, _devices)| user_id.clone())
+            .collect();
```
[the remark this answers is not in the document]
Comment (richvdh) [reply crates/matrix-sdk-crypto/src/identities/manager.rs]: I've rewritten this as a single loop. To be honest I'm not entirely convinced it saves much in the way of memory allocation (we do want to log users_with_no_devices_on_failed_servers, on a single log record), but it's probably clearer.

**Review by bnjbvr — APPROVED** [review]
Great results, thanks!

**Pull request merged**